File: scripts/generate-icons.js
#!/usr/bin/env node
/**
 * Generate PWA icon PNGs from scratch (no dependencies).
 * Run: node scripts/generate-icons.js
 */
const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

function createPNG(width, height, filepath) {
  function pixelColor(x, y, w, h) {
    const cx = w / 2, cy = h / 2;
    const r = Math.min(w, h) / 2;
    const dx = x - cx, dy = y - cy;
    const dist = Math.sqrt(dx * dx + dy * dy);

    if (dist > r) return [0, 0, 0, 0];

    let alpha = 255;
    if (dist > r - 1.5) alpha = Math.round(255 * Math.max(0, (r - dist) / 1.5));

    const t = (x + y) / (w + h);
    let rr = Math.round(255 * (1 - t) + 255 * t);
    let gg = Math.round(139 * (1 - t) + 107 * t);
    let bb = Math.round(85 * (1 - t) + 53 * t);

    const margin = w * 0.22;
    const top = h * 0.25;
    const bottom = h * 0.75;
    const letterW = w - 2 * margin;
    const stroke = letterW * 0.16;
    const lx = x - margin;
    const ly = y - top;
    const lh = bottom - top;

    let isM = false;
    if (y >= top && y <= bottom && x >= margin && x <= w - margin && lh > 0) {
      if (lx <= stroke) isM = true;
      else if (lx >= letterW - stroke) isM = true;
      else if (lx <= letterW / 2) {
        const diagX = (ly / lh) * (letterW / 2);
        if (Math.abs(lx - diagX) < stroke * 0.8) isM = true;
      } else {
        const diagX = letterW - (ly / lh) * (letterW / 2);
        if (Math.abs(lx - diagX) < stroke * 0.8) isM = true;
      }
    }

    if (isM) return [255, 255, 255, alpha];
    return [rr, gg, bb, alpha];
  }

  const raw = Buffer.alloc(height * (1 + width * 4));
  let offset = 0;
  for (let y = 0; y < height; y++) {
    raw[offset++] = 0; // filter byte
    for (let x = 0; x < width; x++) {
      const [r, g, b, a] = pixelColor(x, y, width, height);
      raw[offset++] = r;
      raw[offset++] = g;
      raw[offset++] = b;
      raw[offset++] = a;
    }
  }

  function crc32(buf) {
    let crc = 0xFFFFFFFF;
    for (let i = 0; i < buf.length; i++) {
      crc ^= buf[i];
      for (let j = 0; j < 8; j++) {
        crc = (crc >>> 1) ^ (crc & 1 ? 0xEDB88320 : 0);
      }
    }
    return (crc ^ 0xFFFFFFFF) >>> 0;
  }

  function makeChunk(type, data) {
    const len = Buffer.alloc(4);
    len.writeUInt32BE(data.length);
    const typeData = Buffer.concat([Buffer.from(type), data]);
    const crc = Buffer.alloc(4);
    crc.writeUInt32BE(crc32(typeData));
    return Buffer.concat([len, typeData, crc]);
  }

  const sig = Buffer.from([0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A]);

  const ihdrData = Buffer.alloc(13);
  ihdrData.writeUInt32BE(width, 0);
  ihdrData.writeUInt32BE(height, 4);
  ihdrData[8] = 8;  // bit depth
  ihdrData[9] = 6;  // color type: RGBA
  ihdrData[10] = 0; // compression
  ihdrData[11] = 0; // filter
  ihdrData[12] = 0; // interlace
  const ihdr = makeChunk('IHDR', ihdrData);

  const compressed = zlib.deflateSync(raw, { level: 9 });
  const idat = makeChunk('IDAT', compressed);
  const iend = makeChunk('IEND', Buffer.alloc(0));

  const dir = path.dirname(filepath);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });

  fs.writeFileSync(filepath, Buffer.concat([sig, ihdr, idat, iend]));
  const size = fs.statSync(filepath).size;
  console.log(`Created ${filepath} (${size} bytes)`);
}

const base = path.join(__dirname, '..', 'public', 'icons');
createPNG(192, 192, path.join(base, 'icon-192.png'));
createPNG(512, 512, path.join(base, 'icon-512.png'));
console.log('Done!');

File: scripts/generate-icons.py
#!/usr/bin/env python3
"""Generate PWA icon PNGs from scratch (no dependencies)."""
import struct, zlib, os, math

def create_png(width, height, filepath):
    """Create a PNG icon with orange circle background and white M letter."""
    def pixel_color(x, y, w, h):
        cx, cy = w / 2, h / 2
        r = min(w, h) / 2
        dx, dy = x - cx, y - cy
        dist = math.sqrt(dx * dx + dy * dy)

        if dist > r:
            return (0, 0, 0, 0)

        # Anti-aliasing at edge
        alpha = 255
        if dist > r - 1.5:
            alpha = int(255 * max(0, (r - dist) / 1.5))

        # Orange gradient
        t = (x + y) / (w + h)
        rr = int(255 * (1 - t) + 255 * t)
        gg = int(139 * (1 - t) + 107 * t)
        bb = int(85 * (1 - t) + 53 * t)

        # M letter geometry
        margin = w * 0.22
        top = h * 0.25
        bottom = h * 0.75
        letter_w = w - 2 * margin
        stroke = letter_w * 0.16

        lx = x - margin
        ly = y - top
        lh = bottom - top

        is_m = False
        if top <= y <= bottom and margin <= x <= w - margin and lh > 0:
            if lx <= stroke:
                is_m = True
            elif lx >= letter_w - stroke:
                is_m = True
            elif lx <= letter_w / 2:
                diag_x = (ly / lh) * (letter_w / 2)
                if abs(lx - diag_x) < stroke * 0.8:
                    is_m = True
            else:
                diag_x = letter_w - (ly / lh) * (letter_w / 2)
                if abs(lx - diag_x) < stroke * 0.8:
                    is_m = True

        if is_m:
            return (255, 255, 255, alpha)

        return (rr, gg, bb, alpha)

    raw = bytearray()
    for y in range(height):
        raw.append(0)  # filter byte
        for x in range(width):
            r, g, b, a = pixel_color(x, y, width, height)
            raw.extend([r, g, b, a])

    def chunk(chunk_type, data):
        c = chunk_type + data
        crc = struct.pack('>I', zlib.crc32(c) & 0xFFFFFFFF)
        return struct.pack('>I', len(data)) + c + crc

    sig = b'\x89PNG\r\n\x1a\n'
    ihdr_data = struct.pack('>IIBBBBB', width, height, 8, 6, 0, 0, 0)
    ihdr = chunk(b'IHDR', ihdr_data)
    compressed = zlib.compress(bytes(raw), 9)
    idat = chunk(b'IDAT', compressed)
    iend = chunk(b'IEND', b'')

    os.makedirs(os.path.dirname(filepath), exist_ok=True)
    with open(filepath, 'wb') as f:
        f.write(sig + ihdr + idat + iend)

    print(f'Created {filepath} ({os.path.getsize(filepath)} bytes)')


if __name__ == '__main__':
    base = os.path.join(os.path.dirname(__file__), '..', 'public', 'icons')
    create_png(192, 192, os.path.join(base, 'icon-192.png'))
    create_png(512, 512, os.path.join(base, 'icon-512.png'))
    print('Done!')
